File: android/app/src/main/java/me/nettrash/familyconnect/ui/components/ReportSheet.kt
/*
 * ReportSheet.kt
 * Family Connect (Android)
 *
 * Reporting one member to the family's owner — from a message in the chat,
 * and from a row in the family roster.
 *
 * ONE sheet for both, deliberately. The disclosure it carries is a
 * protocol requirement rather than a nicety ("somebody who reports a
 * message without knowing that has been surprised by their own app"), and
 * two copies of a sentence like that drift apart exactly when it matters.
 */

package me.nettrash.familyconnect.ui.components

import androidx.annotation.StringRes
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.selection.selectableGroup
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.dp
import me.nettrash.familyconnect.R

/**
 * The four reasons a report may carry.
 *
 * The wire strings are the enum's own [wire] values and the protocol's
 * fixed list, so nine locales draw a row from string resources rather than
 * shipping untranslated prose to a moderator (docs/protocol.md,
 * "Reporting a member").
 */
internal enum class ReportReason(val wire: String, @StringRes val label: Int) {
    SPAM("spam", R.string.s_report_reason_spam),
    HARASSMENT("harassment", R.string.s_report_reason_harassment),
    INAPPROPRIATE("inappropriate", R.string.s_report_reason_inappropriate),
    OTHER("other", R.string.s_report_reason_other),
}

/**
 * Report one member to the family's owner.
 *
 * The disclosure is MANDATORY and not a footnote: the owner sees the
 * reporter's name and, when a message was named, its whole text frozen at
 * this moment. Somebody deciding whether to report has to know that before
 * they tap, not afterwards.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ReportSheet(
    displayName: String,
    /**
     * The operator's published contact, or null when unset. The honest
     * escalation path for the case this feature is weakest at: a report
     * ABOUT the owner never reaches the owner (docs/protocol.md,
     * "Reporting a member").
     */
    supportContact: String?,
    onDismiss: () -> Unit,
    onSubmit: (String) -> Unit,
) {
    var reason by remember { mutableStateOf(ReportReason.SPAM) }
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        containerColor = MaterialTheme.colorScheme.surfaceContainerLow,
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 20.dp)
                .padding(bottom = 28.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text(
                text = stringResource(R.string.s_report_member_title, displayName),
                style = MaterialTheme.typography.titleMedium,
            )
            Column(modifier = Modifier.selectableGroup()) {
                ReportReason.entries.forEach { choice ->
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .selectable(
                                selected = reason == choice,
                                role = Role.RadioButton,
                                onClick = { reason = choice },
                            )
                            .padding(vertical = 10.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(12.dp),
                    ) {
                        RadioButton(selected = reason == choice, onClick = null)
                        Text(
                            text = stringResource(choice.label),
                            style = MaterialTheme.typography.bodyLarge,
                        )
                    }
                }
            }
            Text(
                text = stringResource(R.string.s_report_disclosure),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            if (!supportContact.isNullOrEmpty()) {
                HorizontalDivider()
                Text(
                    text = stringResource(R.string.s_if_the_problem_is_the_owner),
                    style = MaterialTheme.typography.labelLarge,
                )
                // VERBATIM, selectable, never linkified: an operator may
                // write an address, a URL or a whole sentence, and three
                // apps guessing differently about which it is would be
                // worse than three apps showing the same text.
                SelectionContainer {
                    Text(text = supportContact, style = MaterialTheme.typography.bodyMedium)
                }
                Text(
                    text = stringResource(R.string.s_operator_published_contact),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.End,
            ) {
                TextButton(onClick = onDismiss) {
                    Text(stringResource(R.string.s_cancel))
                }
                TextButton(onClick = { onSubmit(reason.wire) }) {
                    Text(
                        text = stringResource(R.string.s_report),
                        color = MaterialTheme.colorScheme.error,
                    )
                }
            }
        }
    }
}

File: ios/FamilyConnect/Views/ReportSheet.swift
//
//  ReportSheet.swift
//  FamilyConnect
//
//  Reporting a member to the family owner (docs/protocol.md, "Reporting a
//  member"). Shared by iOS and macOS.
//
//  The four reasons are FIXED, and their raw values are the untranslated
//  wire strings. This is a nine-language product with a nine-language
//  owner: a free-text reason means an inbox the owner cannot read, and a
//  label sent as the reason means a `validation` refusal in whichever
//  language the reporter happens to use.
//

import SwiftUI

/// Which assistant reply is being reported (docs/protocol.md, "Reporting the
/// assistant"). No sender: the assistant is one account and naming it would
/// only invite somebody to pass it to the member-report endpoint, which
/// refuses it.
nonisolated struct AssistantReportTarget: Identifiable, Equatable {
    let messageID: Int64
    /// Whether the reply was in the reporter's own private thread. It changes
    /// nothing about where the report goes — the operator reads both — and is
    /// kept because the sheet says something slightly different about a
    /// reply the family could already read.
    let isPrivateThread: Bool

    var id: Int64 { messageID }
}

/// WHICH SAFETY ROW A BUBBLE GETS, decided in one place for both platforms.
///
/// The two reports are exclusive, and the exclusion is the protocol's: a
/// member report names somebody in your family and the OWNER reads it, while
/// an assistant report names a reply from an account that belongs to no
/// family at all and the OPERATOR reads it. Sending either down the other's
/// endpoint is refused — `not_same_family` one way, `message_not_found` the
/// other — so a menu that offered the wrong one would be a visible failure on
/// a safety screen.
nonisolated enum SafetyRules {
    /// A member's message, acked, not the reader's own and not the
    /// assistant's.
    static func canReportMember(
        senderID: Int64, currentUserID: Int64, isAssistant: Bool, hasServerID: Bool
    ) -> Bool {
        hasServerID && senderID != currentUserID && !isAssistant
    }

    /// An assistant reply, acked. Never the reader's own message, which in
    /// the assistant's own chat is the only other thing there is.
    static func canReportAssistant(
        senderID: Int64, currentUserID: Int64, isAssistant: Bool, hasServerID: Bool
    ) -> Bool {
        hasServerID && senderID != currentUserID && isAssistant
    }
}

/// Who is being reported, and optionally which of their messages.
nonisolated struct ReportTarget: Identifiable, Equatable {
    let senderID: Int64
    let senderName: String
    /// The message named, when the report came from a bubble rather than a
    /// roster row. `nil` reports the PERSON.
    let messageID: Int64?

    var id: Int64 { senderID }
}

/// The protocol's four, and only these.
///
/// The raw value goes on the wire untranslated; `label` is what a person
/// reads. Keeping them apart is what stops a translated string reaching
/// `POST /families/reports`.
nonisolated enum ReportReason: String, CaseIterable, Identifiable {
    case spam
    case harassment
    case inappropriate
    case other

    var id: String { rawValue }

    var label: String {
        switch self {
        case .spam:
            String(localized: "Spam", comment: "Report reason")
        case .harassment:
            String(localized: "Harassment", comment: "Report reason")
        case .inappropriate:
            String(localized: "Inappropriate", comment: "Report reason")
        case .other:
            String(localized: "Something else", comment: "Report reason")
        }
    }
}

struct ReportSheet: View {
    @Environment(AppSession.self) private var session
    let target: ReportTarget
    let onSubmit: (ReportReason) -> Void
    let onCancel: () -> Void

    @State private var reason: ReportReason = .harassment

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    Picker(String(localized: "Reason", comment: "Report sheet: the picker's label"), selection: $reason) {
                        ForEach(ReportReason.allCases) { reason in
                            Text(reason.label).tag(reason)
                        }
                    }
                    #if os(macOS)
                        .pickerStyle(.inline)
                    #endif
                } header: {
                    Text("Why are you reporting \(target.senderName)?",
                         comment: "Report sheet header; %@ is the member's display name")
                } footer: {
                    // MANDATORY, and it is a protocol requirement rather
                    // than a nicety: "somebody who reports a message
                    // without knowing that has been surprised by their own
                    // app". It matters most in a DIRECT chat, where the
                    // owner has never seen the message before.
                    if target.messageID != nil {
                        Text(
                            "Your family owner will see this message and its text.",
                            comment: "Report sheet disclosure")
                    } else {
                        Text(
                            "Your family owner will be told you reported this member.",
                            comment: "Report sheet disclosure")
                    }
                }
                // The honest escalation path for the case this whole
                // feature is weakest at: the moderator IS the owner, so a
                // report about them never reaches them (docs/protocol.md,
                // "Reporting a member"). Absent when the operator has set
                // no contact, and then the section goes with it rather
                // than standing empty.
                if let supportContact = session.supportContact, !supportContact.isEmpty {
                    Section {
                        // VERBATIM, selectable, and never linkified: an
                        // operator may write an address, a URL or a whole
                        // sentence, and three apps guessing differently
                        // about which it is would be worse than three apps
                        // showing the same text.
                        Text(verbatim: supportContact)
                            .textSelection(.enabled)
                    } header: {
                        Text("If the problem is the owner", comment: "Report sheet: the operator's own contact")
                    } footer: {
                        Text(
                            "This server's operator published this contact.",
                            comment: "Report sheet: explains where the support contact came from")
                    }
                }
            }
            .navigationTitle(Text("Report", comment: "Report sheet title and its confirm button"))
            .inlineNavigationTitle()
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel", action: onCancel)
                        .keyboardShortcut(.cancelAction)
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button(String(localized: "Report", comment: "Report sheet title and its confirm button")) { onSubmit(reason) }
                }
            }
        }
        // A Mac sheet cannot be resized by the person using it, so it is
        // sized here or it is wrong for everybody.
        #if os(macOS)
            .frame(width: 420, height: 320)
        #endif
    }
}

/// Reporting an ASSISTANT reply (docs/protocol.md, "Reporting the
/// assistant"). Shared by iOS and macOS.
///
/// Two things make it a different sheet rather than a flag on the one above.
/// It says WHO READS IT — the people who run the server, never the family
/// owner — and that is not a nicety: a private assistant thread belongs to
/// its member alone, so somebody reporting a reply out of one has to know
/// before they send it who is about to see it. And it takes a NOTE, which a
/// member report has none of: the reader here is one operator rather than a
/// nine-language owner, and "it invented a person" is not any of four words.
struct AssistantReportSheet: View {
    @Environment(AppSession.self) private var session
    let target: AssistantReportTarget
    let onSubmit: (ReportReason, String?) -> Void
    let onCancel: () -> Void

    @State private var reason: ReportReason = .inappropriate
    @State private var note: String = ""

    /// The server's cap, enforced here so nobody types past it and is
    /// refused (`POST /reports/assistant`).
    private static let noteLimit = 1000

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    Picker(
                        String(localized: "Reason", comment: "Report sheet: the picker's label"),
                        selection: $reason
                    ) {
                        ForEach(ReportReason.allCases) { reason in
                            Text(reason.label).tag(reason)
                        }
                    }
                    #if os(macOS)
                        .pickerStyle(.inline)
                    #endif
                } header: {
                    Text("What was wrong with this reply?",
                         comment: "Assistant report sheet header")
                }

                Section {
                    TextField(
                        String(localized: "Say something about it (optional)",
                               comment: "Assistant report sheet: the note field"),
                        text: $note, axis: .vertical)
                        .lineLimit(3...6)
                        .onChange(of: note) { _, typed in
                            if typed.count > Self.noteLimit {
                                note = String(typed.prefix(Self.noteLimit))
                            }
                        }
                } footer: {
                    // MANDATORY, and for the reason the member sheet's
                    // disclosure is: somebody who reports a reply out of
                    // their own private thread without knowing who reads it
                    // has been surprised by their own app.
                    Text(
                        "The people who run this server will see this reply and what you write here. Your family will not.",
                        comment: "Assistant report sheet disclosure")
                }

                if let supportContact = session.supportContact, !supportContact.isEmpty {
                    Section {
                        // VERBATIM, selectable, never linkified — the same
                        // rule the member sheet follows.
                        Text(verbatim: supportContact)
                            .textSelection(.enabled)
                    } footer: {
                        Text(
                            "This server's operator published this contact.",
                            comment: "Report sheet: explains where the support contact came from")
                    }
                }
            }
            .navigationTitle(Text("Report this reply", comment: "Assistant report sheet title"))
            .inlineNavigationTitle()
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel", action: onCancel)
                        .keyboardShortcut(.cancelAction)
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button(String(localized: "Report", comment: "Report sheet title and its confirm button")) {
                        let typed = note.trimmingCharacters(in: .whitespacesAndNewlines)
                        onSubmit(reason, typed.isEmpty ? nil : typed)
                    }
                }
            }
        }
        #if os(macOS)
            .frame(minWidth: 420, minHeight: 360)
        #endif
    }
}
